src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -60,15 +86,54 @@ static void WriteAsset(ObjectWriter writer, ISerializerService serializer, Check
             }
         }
 
-        public static ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(Stream stream, int scopeId, ISet<Checksum> checksums, ISerializerService serializerService, CancellationToken cancellationToken)
+        public static ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(PipeReader pipeReader, int scopeId, ISet<Checksum> checksums, ISerializerService serializerService, CancellationToken cancellationToken)
         {
-            using var _ = ArrayBuilder<(Checksum, object)>.GetInstance(out var results);
+            // Workaround for ObjectReader not supporting async reading.
+            // Unless we read from the RPC stream asynchronously and with cancallation support we might hang when the server cancels.
+            // https://github.com/dotnet/roslyn/issues/47861
+
+            // Use local pipe to avoid blocking the current thread on networking IO.
+            var localPipe = new Pipe(PipeOptionsWithUnlimitedWriterBuffer);
+
+            Exception? exception = null;
 
-            using var reader = ObjectReader.TryGetReader(stream, leaveOpen: true, cancellationToken);
+            // start a task on a thread pool thread copying from the RPC pipe to a local pipe:
+            Task.Run(async () =>
+            {
+                try
+                {
+                    await pipeReader.CopyToAsync(localPipe.Writer, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    exception = e;
+                }
+                finally
+                {
+                    await localPipe.Writer.CompleteAsync(exception).ConfigureAwait(false);
+                    await pipeReader.CompleteAsync(exception).ConfigureAwait(false);
```

Comment:
📝 If either of these faults, the resulting exception will not be observed. Should we handle it somehow?